opkg module: allow installing a package in a certain version

joergho opened the pull request:
##### SUMMARY

This extends the community.general.opkg module. It additionally allows to install a package in a certain version.

Tested with opkg 0.4.2 on NI Linux Real-Time 8.8

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
community.general.opkg module

##### ADDITIONAL INFORMATION

opkg module: allow NAME=VERSION syntax to install a package in a certain version

example:
```
- name: Install foo in version 1.2
  community.general.opkg:
    name: foo=1.2
    state: present
```

felixfontein commented on changelogs/fragments/5688-opkg-module-install-certain-version.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - opkg module - allow installing a package in a certain version (https://github.com/ansible-collections/community.general/pull/5688).
```
```suggestion
  - opkg - allow installing a package in a certain version (https://github.com/ansible-collections/community.general/pull/5688).
```

joergho replied:
Changed in [a578f43](https://github.com/ansible-collections/community.general/pull/5688/commits/a578f43459cb809105d4ccdf8b817c3f440e39b5)

felixfontein commented on plugins/modules/opkg.py:
```diff
@@ -22,6 +22,8 @@
     name:
         description:
             - Name of package(s) to install/remove.
+            - NAME=VERSION syntax is also supported to install a package
```
```suggestion
            - C(NAME=VERSION) syntax is also supported to install a package
```

joergho replied:
Changed in [a578f43](https://github.com/ansible-collections/community.general/pull/5688/commits/a578f43459cb809105d4ccdf8b817c3f440e39b5)

felixfontein commented on plugins/modules/opkg.py:
```diff
@@ -22,6 +22,8 @@
         description:
             - Name of package(s) to install/remove.
+            - NAME=VERSION syntax is also supported to install a package
+              in a certain version, see examples
```
```suggestion
              in a certain version. See the examples. This is supported since
              community.general 6.2.0.
```

joergho replied:
Changed in [a578f43](https://github.com/ansible-collections/community.general/pull/5688/commits/a578f43459cb809105d4ccdf8b817c3f440e39b5)

felixfontein commented on plugins/modules/opkg.py:
```diff
@@ -102,16 +109,36 @@ def update_package_db(module, opkg_path):
+        # variable out is one line if the package is installed:
+        # "NAME - VERSION - DESCRIPTION"
+        if version is not None:
+            if not out.startswith("%s - %s" % (name, version)):
```
Shouldn't you then use
```suggestion
            if not out.startswith("%s - %s " % (name, version)):
```
so that `foo - 1.10` does not match when checking for `foo=1.1`?

joergho replied:
Good catch! Changed in [a578f43](https://github.com/ansible-collections/community.general/pull/5688/commits/a578f43459cb809105d4ccdf8b817c3f440e39b5)

felixfontein commented on plugins/modules/opkg.py:
```diff
@@ -102,16 +109,36 @@ def update_package_db(module, opkg_path):
+            if not out.startswith("%s - %s" % (name, version)):
+                return False
+        else:
+            if not out.startswith(name):
```
```suggestion
            if not out.startswith(name + " "):
```
So that `foobar - 1.0` does not match `foo`.

joergho replied:
This is technically not necessary as `opkg list-installed PACKAGE` will only provide information for the package `PACKAGE` if it is installed. But also, it does not hurt. Changed in [a578f43](https://github.com/ansible-collections/community.general/pull/5688/commits/a578f43459cb809105d4ccdf8b817c3f440e39b5)

felixfontein commented on plugins/modules/opkg.py:
```diff
@@ -102,16 +109,36 @@ def update_package_db(module, opkg_path):
-            return True
-
-        return False
+        rc, out, err = module.run_command("%s list-installed %s" % (opkg_path, name))
```
Better use
```suggestion
        rc, out, err = module.run_command([opkg_path, "list-installed", name])
```
Otherwise you need to use `shlex_quote` to avoid the shell parsing this wrongly.

joergho replied:
Changed in [beaf2a1](https://github.com/ansible-collections/community.general/pull/5688/commits/beaf2a186bed1eda88a8f7c1243dc2be4aba6290)

felixfontein commented on plugins/modules/opkg.py:
```diff
@@ -102,16 +109,36 @@ def update_package_db(module, opkg_path):
+
+def split_name_and_version(module, package):
+    """ Split the name and the version when using the NAME=VERSION syntax """
+    splitted = package.split('=')
```
Why not simply
```suggestion
    splitted = package.split('=', 1)
```
and leave away the error check?

joergho replied:
Changed in [a578f43](https://github.com/ansible-collections/community.general/pull/5688/commits/a578f43459cb809105d4ccdf8b817c3f440e39b5)

felixfontein commented on plugins/modules/opkg.py:
```diff
@@ -154,13 +183,20 @@ def install_packages(module, opkg_path, packages):
 
-        if not query_package(module, opkg_path, package):
-            module.fail_json(msg="failed to install %s: %s" % (package, out))
+        rc, out, err = module.run_command("%s install %s %s%s" % (opkg_path, force, package, version_str))
```
Better use a list here as well.

joergho replied:
Changed in [beaf2a1](https://github.com/ansible-collections/community.general/pull/5688/commits/beaf2a186bed1eda88a8f7c1243dc2be4aba6290)

felixfontein reviewed (commented):
Thanks for your contribution! I've added some first comments.

joergho commented:
@felixfontein
Thanks for your review. I addressed all your review comments. 

I also noticed that there was a bug that parameter `force=reinstall` would not actually reinstall a package if it was already installed. I introduced a fix for this: [5bd3977](https://github.com/ansible-collections/community.general/pull/5688/commits/5bd3977ce742f8a2c731f066cbceec549544b5bf)

felixfontein commented:
Could you please move the bugfix to a new PR so we can backport it also to stable-5? Thanks.

joergho commented:
> Could you please move the bugfix to a new PR so we can backport it also to stable-5? Thanks.

Sure. See PR #5705

felixfontein reviewed (commented):
Looks good to me (looking at the code; I cannot test it). Now there is a conflict that needs to be resolved though (please rebase).

joergho commented:
> Looks good to me (looking at the code; I cannot test it). Now there is a conflict that needs to be resolved though (please rebase).

Rebased in [5232803](https://github.com/ansible-collections/community.general/pull/5688/commits/523280399db1eb5c4debaad6dc2f04bd842da7f6)

Also I executed a test with python3.10 and opkg 0.4.2 on NI Linux Real-Time 8.8:

opkg-tests.yml:
```yaml
- name: "opkg tests"
  gather_facts: no

  tasks:
  - name: "opkg remove (1)"
    community.general.opkg:
        name: "tmux"
        state: "absent"

  - name: "opkg remove (2)"
    community.general.opkg:
        name: "tmux"
        state: "absent"

  - name: "opkg install (1)"
    community.general.opkg:
        name: "tmux"
        state: "present"

  - name: "opkg install (2)"
    community.general.opkg:
        name: "tmux"
        state: "present"

  - name: "install --force-reinstall"
    community.general.opkg:
        name: "tmux"
        state: "present"
        force: "reinstall"

  - name: "opkg remove (3)"
    community.general.opkg:
        name: "tmux"
        state: "absent"

  - name: "opkg install with version"
    community.general.opkg:
        name: "tmux=2.6-r0.18"
        state: "present"

  - name: "opkg install with incorrect version"
    community.general.opkg:
        name: "tmux=2.6"
        state: "present"
```

$ ansible-playbook opkg-tests.yml

PLAY [opkg tests] **************************************************************************************************************************

TASK [opkg remove (1)] **************************************************************************************************************************
changed: [agent01]

TASK [opkg remove (2)] **************************************************************************************************************************
ok: [agent01]

TASK [opkg install (1)] **************************************************************************************************************************
changed: [agent01]

TASK [opkg install (2)] **************************************************************************************************************************
ok: [agent01]

TASK [install --force-reinstall] **************************************************************************************************************************
changed: [agent01]

TASK [opkg remove (3)] **************************************************************************************************************************
changed: [agent01]

TASK [opkg install with version] **************************************************************************************************************************
changed: [agent01]

TASK [opkg install with incorrect version] **************************************************************************************************************************
fatal: [agent01]: FAILED! => {"changed": false, "msg": "failed to install tmux=2.6: No packages installed or removed.\n"}

PLAY RECAP **************************************************************************************************************************
agent01 : ok=7    changed=5    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

felixfontein commented:
Thanks for testing it! :) I'll keep this open a few more days and will then merge it if nobody objects until then.

felixfontein commented:
Merging so that #5718 can rebase and advance.

felixfontein commented:
@joergho thanks for contributing this!
